Source/MultipartFormData.swift

```diff
@@ -0,0 +1,624 @@
+
+            completionHandler(error)
+            return
+        }
+
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)) {
+            outputStream.scheduleInRunLoop(NSRunLoop.currentRunLoop(), forMode: NSDefaultRunLoopMode)
+            outputStream.open()
+
+            self.bodyParts.first?.hasInitialBoundary = true
+            self.bodyParts.last?.hasFinalBoundary = true
+
+            var error: NSError?
+
+            for bodyPart in self.bodyParts {
+                if let writeError = self.writeBodyPart(bodyPart, toOutputStream: outputStream) {
+                    error = writeError
+                    break
+                }
+            }
+
+            outputStream.close()
+            outputStream.removeFromRunLoop(NSRunLoop.currentRunLoop(), forMode: NSDefaultRunLoopMode)
+
+            dispatch_async(dispatch_get_main_queue()) {
+                completionHandler(error)
+            }
+        }
+    }
+
+    // MARK: - Private - Body Part Encoding
+
+    private func encodeBodyPart(bodyPart: BodyPart) -> EncodingResult {
+        let encoded = NSMutableData()
+
+        let initialData = bodyPart.hasInitialBoundary ? initialBoundaryData() : encapsulatedBoundaryData()
+        encoded.appendData(initialData)
+
+        let headerData = encodeHeaderDataForBodyPart(bodyPart)
+        encoded.appendData(headerData)
+
+        let bodyStreamResult = encodeBodyStreamDataForBodyPart(bodyPart)
+
+        switch bodyStreamResult {
+        case .Failure:
+            return bodyStreamResult
+        case let .Success(data):
+            encoded.appendData(data)
+        }
+
+        if bodyPart.hasFinalBoundary {
+            encoded.appendData(finalBoundaryData())
+        }
+
+        return .Success(encoded)
+    }
+
+    private func encodeHeaderDataForBodyPart(bodyPart: BodyPart) -> NSData {
+        var headerText = ""
+
+        for (key, value) in bodyPart.headers {
+            headerText += "\(key): \(value)\(EncodingCharacters.CRLF)"
+        }
+        headerText += EncodingCharacters.CRLF
+
+        return headerText.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)!
+    }
+
+    private func encodeBodyStreamDataForBodyPart(bodyPart: BodyPart) -> EncodingResult {
+        let inputStream = bodyPart.bodyStream
+        inputStream.scheduleInRunLoop(NSRunLoop.currentRunLoop(), forMode: NSDefaultRunLoopMode)
+        inputStream.open()
+
+        var error: NSError?
+        let encoded = NSMutableData()
+
+        while inputStream.hasBytesAvailable {
+            var buffer = [UInt8](count: self.streamBufferSize, repeatedValue: 0)
+            let bytesRead = inputStream.read(&buffer, maxLength: self.streamBufferSize)
+
+            if inputStream.streamError != nil {
+                error = inputStream.streamError
+                break
+            }
+
+            if bytesRead < 0 {
```

Comment:
That's a great catch @kylef. I agree completely. I'll get this method fixed up and also apply this approach to the other streaming methods as well.